### DISCUSSION TITLE
How do I get &AppHandle instance in “extern "system" fn”so that I can emit message to js?

### DISCUSSION THREAD

**Discussion by Kalmaegi** [tauri-apps/tauri#6309] [Q&A]
I'm using windows api that listens to mouse and keyboard, like "SetWindowsHookExA" but this api needs a function address for callbacks, is there any way I can get AppHandle in one function? If there is a global reference to it, I searched the documentation but couldn't find the corresponding method, thanks a lot

**Comment by FabianLars** [comment]
I'd suggest looking into the [`once_cell`](https://crates.io/crates/once_cell) crate. You could move an AppHandle from tauri's setup hook (the earliest point in time where you can get an AppHandle) into a `OnceCell` and access it similar to a global variable.

**Comment by wtto00** [comment]
I use [once_cell](https://crates.io/crates/once_cell) like this:
```rust
// Declare global static variable
static APP_HANDLE: OnceCell<Mutex<AppHandle>> = OnceCell::new();

// Set APP_HANDLE
APP_HANDLE
  .set(Mutex::new(app.app_handle().to_owned()))
  .unwrap();

// Get APP_HANDLE
let app = APP_HANDLE.get().unwrap().lock();
if let Ok(app_handle) = app {
    app_handle
        .emit_all("plugin:iap|exception", ExceptionError { code, message })
        .unwrap();
}
```
To set `APP_HANDLE`, I had to remove `<R: Runtime>` in `pub fn init<R: Runtime>() -> TauriPlugin<R>`, and changed it to `pub fn init() -> TauriPlugin<Wry>`.

**Comment by merlinaudio** [comment]
From the other comments, here's how I did it with Tauri v2

```rust
use std::sync::OnceLock; // OnceLock is part of std now

static APP_HANDLE: OnceLock<AppHandle> = OnceLock::new();

#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    let context = tauri::generate_context!();

    let app = tauri::Builder::default()
        // config omitted for brevity
        .build(context)
        .unwrap();

    APP_HANDLE.set(app.app_handle().to_owned()).unwrap();

    app.run(|_, _| {});
}
```

Then I made a small helper function to use it:

```rust
fn app_handle<'a>() -> &'a AppHandle {
    APP_HANDLE.get().unwrap()
}

// Enjoy:
fn foo() {
  let app_handle = app_handle();
}
```

**Comment by Z3NTL3** [comment]
Hi to update,

As @merlinaudio said, ``OnceLock`` is a better viable option in my opinion as:

It's the Sync variant of  OnceCell from std:
> The corresponding [Sync](https://doc.rust-lang.org/std/marker/trait.Sync.html) version of OnceCell<T> is [OnceLock<T>](https://doc.rust-lang.org/std/sync/struct.OnceLock.html).

I just do

```rust
use std::sync::OnceLock;

static APP_HANDLE: OnceLock<AppHandle> = OnceLock::new();

// and in my Tauri .setup func:

 APP_HANDLE.set(app.app_handle().to_owned()).unwrap();

```

**Reply by FabianLars** [reply]
Yep, nowadays the std version makes more sense than the once_cell crate. I'll collapse my comment as outdated.
